README.md
<h3 align="center">
  kong-plugin-proxy-cache
</h3>

<p align="center">
    A Proxy Caching plugin for <a href="https://konghq.com/">Kong</a>
</p>

<p align="center">
  <a href="./LICENSE"><img src="https://img.shields.io/github/license/mashape/apistatus.svg?style=flat-square"></a>
  <a href="https://luarocks.org/modules/sergiojorge/kong-plugin-proxy-cache"><img src="https://img.shields.io/luarocks/v/sergiojorge/kong-plugin-proxy-cache.svg?style=flat-square"></a>
  <a href="https://www.globo.com/"><img src="https://img.shields.io/badge/powered%20by-globo.com-blue.svg?style=flat-square"></a>
</p>

A Proxy Caching plugin for Kong makes it fast and easy to configure caching of responses and serving of those cached responses in Redis. It caches responses bases on configurable response code and request headers with the request method.

This Kong plugin adds a non-standard `X-Cache-Status` header. There are several possible values for this header:

* `MISS`: The resource was not found in cache, and the request was proxied upstream.
* `HIT`: The request was satisfied and served from cache.
* `BYPASS`: The request could not be satisfied from cache based on plugin settings.

## Getting started

Configure this plugin on a Service by making the following request:

```shell
$ curl -X POST http://kong:8001/services/debug-upstream/plugins \
    --data "name=proxy-cache" \
    --data "config.cache_ttl=300" \
    --data "config.redis.host=127.0.0.1"
```

* **service**: the id or name of the Service that this plugin configuration will target.

## Configuration

Here's a list of all the settings which can be used in this plugin:

> Note: The required fields are in bold.

| Field          | Default       | Description
|----------------|---------------|----------------------------------------------------
| **response_code**  | 200, 301, 302 | Upstream response status code considered cacheable
| vary_headers   |               | Relevant headers considered for the cache key
| vary_nginx_variables |              | Relevant nginx variables considered for the cache key
| **cache_ttl**      | 300           | TTL, in seconds, of cache responses
| cache_control  | true         | Respect the Cache-Control behaviors
| **redis.host**     |               | Host to use for Redis connection
| **redis.port**     | 6379          | Port to use for Redis connection
| **redis.timeout**  | 2000          | Connection timeout to use for Redis connection
| redis.password |               | Password to use for Redis connection
| **redis.database** | 0             | Database to use for Redis connection
| redis.sentinel_master_name  |           | Sentinel master name to use for Redis connection. Defining this value implies using Redis Sentinel
| redis.sentinel_role         |           | Sentinel role to use for Redis connection
| redis.sentinel_addresses    |           | Sentinel address to use for Redis connection
| **redis.max_idle_timeout** | 10000       | maximal idle timeout (in ms) for keep alive the Redis connection
| **redis.pool_size** | 1000       | maximal pool size by Redis connection

### Cache Control

When the `cache_control` is enabled by settings, Kong will respect request and response `Cache-Control` headers as defined by RFC7234, with a few notes:

* The behavior of no-cache is simplified to exclude the entity from being cached entirely.
* Secondary key calculation via Vary is not yet supported.

## Installing

### From LuaRocks

```shell
$ luarocks install kong-plugin-proxy-cache
```

### From GitHub

```shell
$ sh -c "git clone https://github.com/globocom/kong-plugin-proxy-cache /tmp/kong-plugin-proxy-cache && cd /tmp/kong-plugin-proxy-cache && luarocks make *.rockspec"
```

## Contributing

Please, read the contribute guide [CONTRIBUTING](./CONTRIBUTING.md).

## License

kong-plugin-proxy-cache is [MIT licensed](./LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

- Changes that have landed in master but are not yet released.

## 2.0.0 - 2018-11-13
### Added
- Add Redis Sentinel support

### Changed
- Change the storage to use `pintsized/lua-resty-redis-connector` instead of `openresty/lua-resty-redis`

### Fixed
- Fix error `API disabled in the current context` when `Cache-Control` is `true`

## 1.3.3 - 2018-11-12
### Fixed
- Fix error `API disabled in the current context` when `Cache-Control` is `true`

## 1.3.2 - 2018-11-07
### Changed
- Change the default value for `cache_control` because the RFC behavior must be enabled by default

## 1.3.1 - 2018-10-31
### Fixed
- Change the body_filter should not set the `X-Cache-Status` header.

## 1.3.0 - 2018-10-30
### Added
- Add `max_idle_timeout` and `pool_size` as schema
- Add filter to remove hop-by-hop headers

### Changed
- Improves performance

## 1.2.5 - 2018-10-29
### Changed
- Improves performance

## 1.2.4 - 2018-10-22
### Changed
- Improves performance

### Fixed
- The plugin should not process the `:body_filter(config)` when `rt_body_chunks` is `nil`.

## 1.2.3 - 2018-10-22
### Changed
- Improves performance

## 1.2.2 - 2018-10-22
### Changed
- Improves performance

## 1.2.1 - 2018-10-17
### Changed
- Improves performance

## 1.2.0 - 2018-10-15
### Fixed
- The `Cache-Control` was implemented incorrectly. It was respecting the client header instead of the upstream header.

### Changed
- Change default `response_code` to `200`, `301` and `302` like the nginx default config.

### Removed
- Remove `REFRESH` from `X-Cache-Status`.
- Remove `Cache-Control: no-cache` validation on access.

## 1.1.0 - 2018-10-09
### Added
- The `ngx.host` was added to compose cache key.

### Fixed
- The cache key composition considered only the last nginx variable.

## 1.0.1 - 2018-10-04
### Added
- Add error handling on plugin exceptions.

## 1.0.0 - 2018-10-03
### Added
- This CHANGELOG file.

[1.3.2]: https://github.com/globocom/kong-plugin-proxy-cache/compare/1.3.1...1.3.2
[1.3.1]: https://github.com/globocom/kong-plugin-proxy-cache/compare/1.3.0...1.3.1
[1.3.0]: https://github.com/globocom/kong-plugin-proxy-cache/compare/1.2.5...1.3.0
[1.2.5]: https://github.com/globocom/kong-plugin-proxy-cache/compare/1.2.4...1.2.5
[1.2.4]: https://github.com/globocom/kong-plugin-proxy-cache/compare/1.2.3...1.2.4
[1.2.3]: https://github.com/globocom/kong-plugin-proxy-cache/compare/1.2.2...1.2.3
[1.2.2]: https://github.com/globocom/kong-plugin-proxy-cache/compare/1.2.1...1.2.2
[1.2.1]: https://github.com/globocom/kong-plugin-proxy-cache/compare/1.2.0...1.2.1
[1.2.0]: https://github.com/globocom/kong-plugin-proxy-cache/compare/1.1.0...1.2.0
[1.1.0]: https://github.com/globocom/kong-plugin-proxy-cache/compare/1.0.1...1.1.0
[1.0.1]: https://github.com/globocom/kong-plugin-proxy-cache/compare/1.0.0...1.0.1


proxy-cache/schema.lua
local function server_port(given_value, given_config)
    if given_value > 65534 then
        return false, "port value too high"
    end
end

return {
    no_consume = true,

    fields = {
        response_code = {
            type = "array",
            default = {"200", "301", "302"},
            required = true
        },
        vary_headers = {
            type = "array",
            required = false
        },
        vary_nginx_variables = {
            type = "array",
            required = false
        },
        cache_ttl = {
            type = "number",
            default = 300,
            required = true
        },
        cache_control = {type = "boolean", default = true},
        redis = {
            type = "table",
            schema = {
                fields = {
                    host = {type = "string", required = false},
                    sentinel_master_name = {type = "string", required = false},
                    sentinel_role = {type = "string", required = false, default = "master"},
                    sentinel_addresses = {type = "array", required = false},
                    port = {
                        type = "number",
                        func = server_port,
                        default = 6379,
                        required = true
                    },
                    timeout = {type = "number", required = true, default = 2000},
                    password = {type = "string", required = false},
                    database = {type = "number", required = true, default = 0},
                    max_idle_timeout = {type = "number", required = true, default = 10000},
                    pool_size = {type = "number", required = true, default = 1000}
                }
            }
        }
    }
}
